File: 139.Word_Break.java
// recursion DP
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        // 可不用新存 hashset，只为更快速判断is curr word exist in wordDict
        Set<String> dict = new HashSet<String>(wordDict);
        // 存储看过的sub string true/false
        Map<String, Boolean> mem = new HashMap<String, Boolean>();
        return wordBreakHelper(dict, mem, s);
    }
    private boolean wordBreakHelper(Set<String> dict, Map<String, Boolean> mem, String curr) {
        // two end conditions
        if (dict.contains(curr)) {
            mem.put(curr, true);
            return true;
        }
        if (mem.containsKey(curr)) return mem.get(curr);
        
        for (int i=1; i<curr.length(); ++i) {
            String left = curr.substring(0,i);
            String right = curr.substring(i);
            // 重点：如果一部分在dict里，另一部分call recursion returned true， 直接true
            if (dict.contains(right) && wordBreakHelper(dict,mem,left)) {
                mem.put(curr, true);
                return true;
            }
        }
        mem.put(curr, false);
        return false;
    }
}

File: 139.WordBreak.py
class Solution:
    '''
    dp[i]: true/false until the ith character in s
    transform function: dp[i] = dp[i] or (dp[j] and s[j:i] in wordDict)
        - 如果s[j:i]存在dict中，并且dp[j]==True（s[0:j]也满足分割条件）, 那么dp[i] 也满足分割条件
    init: dp[0] = True, other False
    traversal: two-loop 
        1. traverse s: range(0, len(s)+1)
        2. traverse sub-string range(0, i)
    '''
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False] * (n+1)
        dp[0] = True
        for i in range(n+1):
            for j in range(i):
                dp[i] = dp[i] or (dp[j] and self.checkDict(s[j: i], wordDict))
        return dp[n]
    
    def checkDict(self, s, wordDict):
        if s in wordDict:
            return True
        return False